pkg/compact/downsample/streamed_block_writer.go

```diff
@@ -0,0 +1,296 @@
+	"github.com/prometheus/tsdb/fileutil"
+	"github.com/prometheus/tsdb/index"
+	"github.com/prometheus/tsdb/labels"
+)
+
+type labelValues map[string]struct{}
+
+func (lv labelValues) add(value string) {
+	lv[value] = struct{}{}
+}
+
+func (lv labelValues) get(set *[]string) {
+	for value := range lv {
+		*set = append(*set, value)
+	}
+}
+
+type labelsValues map[string]labelValues
+
+func (lv labelsValues) add(labelSet labels.Labels) {
+	for _, label := range labelSet {
+		values, ok := lv[label.Name]
+		if !ok {
+			// Add new label.
+			values = labelValues{}
+			lv[label.Name] = values
+		}
+		values.add(label.Value)
+	}
+}
+
+// streamedBlockWriter writes downsampled blocks to a new data block. Implemented to save memory consumption
+// by means writing chunks data right into the files, omitting keeping them in-memory. Index and meta data should be
+// sealed afterwards, when there aren't more series to process.
+type streamedBlockWriter struct {
+	blockDir     string
+	finalized    bool // set to true, if Close was called
+	logger       log.Logger
+	interrupt    bool // interrupt says Close no to finalize block due to internal error.
+	meta         metadata.Meta
+	totalChunks  uint64
+	totalSamples uint64
+
+	chunkWriter tsdb.ChunkWriter
+	indexWriter tsdb.IndexWriter
+	indexReader tsdb.IndexReader
+	closers     []io.Closer
+
+	labelsValues labelsValues       // labelsValues list of used label sets: name -> []values.
+	memPostings  *index.MemPostings // memPostings contains references from label name:value -> postings.
+	postings     uint64             // postings is a current posting position.
+}
+
+// NewStreamedBlockWriter returns streamedBlockWriter instance, it's not concurrency safe.
+// Caller is responsible to Close all io.Closers by calling the Close when downsampling is done.
+// In case if error happens outside of the StreamedBlockWriter during the processing,
+// index and meta files will be written anyway, so the caller is always responsible for removing block directory with
+// a garbage on error.
+// This approach simplifies StreamedBlockWriter interface, which is a best trade-off taking into account the error is an
+// exception, not a general case.
+func NewStreamedBlockWriter(
+	blockDir string,
+	indexReader tsdb.IndexReader,
+	logger log.Logger,
+	originMeta metadata.Meta,
+) (sw *streamedBlockWriter, err error) {
+	sw = &streamedBlockWriter{
+		logger:       logger,
+		blockDir:     blockDir,
+		indexReader:  indexReader,
+		meta:         originMeta,
+		closers:      make([]io.Closer, 0, 2),
+		labelsValues: make(labelsValues, 1024),
+		memPostings:  index.NewUnorderedMemPostings(),
+	}
+
+	// We should close any opened Closer up to an error.
+	defer func() {
+		if err != nil {
+			sw.interrupt = true
+			var merr tsdb.MultiError
+			merr.Add(err)
+			merr.Add(sw.Close())
+			err = merr.Err()
+		}
+	}()
+
+	sw.chunkWriter, err = chunks.NewWriter(filepath.Join(blockDir, block.ChunksDirname))
```

Comment:
I have an idea to pass writers as a dependency, but in that case NewStreamedBlockWriter wouldn't be responsible for block's consistency. any thoughts? @domgreen @bwplotka

Reply:
Good point, was thinking about this as well.

so if we pass it as arguments, it will make testing easier, however by doing it like this we make streamwriter responsible for all writers making out interface better and with deep functionality, so I think it's fine for now. :+1: